File: 프로그래머스/3/42898. 등굣길/등굣길.js
function solution(m, n, puddles) {
    const arr = Array.from({length: n}, () => Array(m).fill(0));
    const MOD = 1000000007;
    
    for (let [x, y] of puddles) {
        arr[y-1][x-1] = -1;
    }
    
    if (arr[0][0] !== -1) arr[0][0] = 1;
    
    for (let r = 0; r < n; r++) {
        for (let c = 0; c < m; c++) {
            if (arr[r][c] === -1) {
                arr[r][c] = 0;
                continue;
            }
            if (r > 0) arr[r][c] += arr[r-1][c];
            if (c > 0) arr[r][c] += arr[r][c-1];
            arr[r][c] %= MOD;
        }
    }
    
    return arr[n-1][m-1];
}

File: 프로그래머스/3/42898. 등굣길/등굣길.py
def solution(m, n, puddles):
    dp = [[0] * m for _ in range(n)] 
    dp[0][0] = 1
    
    puddles = set((y-1, x-1) for [x, y] in puddles)
    MOD = 1_000_000_007
    
    for r in range(n):
        for c in range(m):
            if (r, c) in puddles:
                continue
                
            if r == 0 and c == 0:
                continue
            elif r == 0:    
                dp[r][c] = dp[r][c-1] % MOD
            elif c == 0:
                dp[r][c] = dp[r-1][c] % MOD
            else:
                dp[r][c] = (dp[r-1][c] + dp[r][c-1]) % MOD
                
    return dp[n-1][m-1]